libminifi/src/core/ProcessSession.cpp

```diff
@@ -540,111 +481,97 @@ void ProcessSession::import(const std::string& source, std::vector<std::shared_p
+        if (claim == nullptr) {
+          startTime = getTimeMillis();
+          claim = std::make_shared<ResourceClaim>(process_context_->getContentRepository());
+        }
+        if (stream == nullptr) {
+          stream = process_context_->getContentRepository()->write(claim);
+        }
+        if (stream == nullptr) {
+          logger_->log_error("Stream is null");
+          rollback();
+          return;
+        }
+        if (stream->write(begin, len) != len) {
+          logger_->log_error("Error while writing");
           stream->closeStream();
-          std::string details = process_context_->getProcessorNode()->getName() + " modify flow record content " + flowFile->getUUIDStr();
-          uint64_t endTime = getTimeMillis();
-          provenance_report_->modifyContent(flowFile, details, endTime - startTime);
-          flows.push_back(flowFile);
-
-          /* Reset these to start processing the next FlowFile with a clean slate */
-          flowFile.reset();
-          stream.reset();
-          claim.reset();
-
-          /* Skip delimiter */
-          begin = delimiterPos + 1;
+          throw Exception(FILE_OPERATION_EXCEPTION, "File Export Error creating Flowfile");
+        }
+
+        /* Create a FlowFile if we reached a delimiter */
+        if (delimiterPos == end) {
+          break;
         }
+        flowFile = std::static_pointer_cast<FlowFileRecord>(create());
+        flowFile->setSize(stream->getSize());
+        flowFile->setOffset(0);
+        flowFile->setResourceClaim(claim);
+        logging::LOG_DEBUG(logger_) << "Import offset " << flowFile->getOffset() << " length " << flowFile->getSize() << " content " << flowFile->getResourceClaim()->getContentFullPath()
+                                    << ", FlowFile UUID " << flowFile->getUUIDStr();
+        stream->closeStream();
+        std::string details = process_context_->getProcessorNode()->getName() + " modify flow record content " + flowFile->getUUIDStr();
+        uint64_t endTime = getTimeMillis();
+        provenance_report_->modifyContent(flowFile, details, endTime - startTime);
+        flows.push_back(flowFile);
+
+        /* Reset these to start processing the next FlowFile with a clean slate */
+        flowFile.reset();
+        stream.reset();
+        claim.reset();
+
+        /* Skip delimiter */
+        begin = delimiterPos + 1;
       }
-    } catch (std::exception &exception) {
-      logger_->log_debug("Caught Exception %s", exception.what());
-      throw;
-    } catch (...) {
-      logger_->log_debug("Caught Exception during process session write");
-      throw;
     }
+  } catch (std::exception &exception) {
+    logger_->log_debug("Caught Exception %s", exception.what());
+    throw;
   } catch (...) {
-    if (flowFile != nullptr && claim != nullptr && flowFile->getResourceClaim() == claim) {
-      flowFile->getResourceClaim()->decreaseFlowFileRecordOwnedCount();
-      flowFile->clearResourceClaim();
-    }
```

Comment:
I think we still need to clean up the resource claim (`clearResourceClaim()`) of the flow file before proceeding, because the flow file destructor will not do so. Is this done somewhere else?

Reply:
the whole point of this story is that once the FlowFile assumes ownership over a ResourceClaim it should clean up after itself, currently this cleanup is in the `FlowFileRecord::~FlowFileRecord` which is arguably not the best place but only a `FlowFileRecord` can notify the content repository

Reply:
I just now realized that `ResourceClaim` also holds a reference to a `StreamManager<ResourceClaim>` it is beyond me why the `ResourceClaim` doesn't clean up after itself

Reply:
before `1.0.0` we should migrate the cleanup logic to `ResourceClaim`, currently it is not feasible as all methods in the `StreamManager` expect a `shared_ptr<ResourceClaim>` and neither in the constructor nor in the destructor do we have the means to summon one 😞

Reply:
I don't think that the `content_repo_` in the `FlowFileRecord` is different from the one in `ResourceClaim`. I agree with the plans to migrate the cleanup logic to `ResourceClaim`, finally giving us complete and encapsulated RAII semantics.

We have the means to summon a `std::shared_ptr<ResourceClaim>` from `ResourceClaim::~ResourceClaim` by calling `shared_from_this()`.

Reply:
unfortunately we do not, because it is most probably destroyed through a `shared_ptr` which first decrements the counter, then calls the destructor and with `shared_ptr` we cannot increment if the counter reached 0 (as it just did because that triggers the destructor) and we get a nice `bad_weak_ptr`

Reply:
good point